fdbserver/ptxn/MessageSerializer.cpp

```diff
@@ -0,0 +1,348 @@
+
+	DeserializerImpl deserializer(serializedArena, serialized);
+	header = deserializer.deserializeAsMainHeader();
+
+	endIterator = iterator(serializedArena, serialized, true);
+}
+
+const StorageTeamID& SubsequencedMessageDeserializer::getStorageTeamID() const {
+	return header.storageTeamID;
+}
+
+size_t SubsequencedMessageDeserializer::getNumVersions() const {
+	return header.numItems;
+}
+
+const Version& SubsequencedMessageDeserializer::getFirstVersion() const {
+	return header.firstVersion;
+}
+
+const Version& SubsequencedMessageDeserializer::getLastVersion() const {
+	return header.lastVersion;
+}
+SubsequencedMessageDeserializer::iterator::iterator(const Arena& serializedArena_,
+                                                    StringRef serialized_,
+                                                    bool isEndIterator)
+  : deserializer(serializedArena_, serialized_), rawSerializedData(serialized_) {
+
+	header = deserializer.deserializeAsMainHeader();
+
+	// Set the iterator to a state that has consumed all items in the current version section
+	// See TLogStorageServerMessageTLogStorageServerMessageDeserializer::begin
+	sectionIndex = -1;
+	versionHeader.numItems = 0;
+	itemIndex = 0;
+
+	// If it is marked as end, place the iterator to the end
+	if (isEndIterator) {
+		sectionIndex = header.numItems;
+	}
+}
+
+bool SubsequencedMessageDeserializer::iterator::operator==(const iterator& another) const {
+	return rawSerializedData.begin() == another.rawSerializedData.begin() &&
+	       rawSerializedData.size() == another.rawSerializedData.size() && sectionIndex == another.sectionIndex &&
+	       itemIndex == another.itemIndex;
+}
+
+bool SubsequencedMessageDeserializer::iterator::operator!=(const iterator& another) const {
+	return !(*this == another);
+}
+
+SubsequencedMessageDeserializer::iterator::reference SubsequencedMessageDeserializer::iterator::operator*() const {
+	return currentItem;
+}
+
+SubsequencedMessageDeserializer::iterator::pointer SubsequencedMessageDeserializer::iterator::operator->() const {
+	return &currentItem;
+}
+
+SubsequencedMessageDeserializer::iterator& SubsequencedMessageDeserializer::iterator::operator++() {
+	// Move to next section if the current section is consumed
+	if (itemIndex == versionHeader.numItems) {
+		++sectionIndex;
+		itemIndex = 0;
+		while (sectionIndex != header.numItems) {
+			versionHeader = deserializer.deserializeAsSectionHeader();
+			currentItem.version = versionHeader.version;
+			if (versionHeader.numItems != 0) {
+				break;
+			}
+			++sectionIndex;
+		}
+
+		if (sectionIndex == header.numItems) {
+			// Reached the end, do not try to deserialize the data
+			ASSERT(deserializer.allConsumed());
+			return *this;
+		}
+	}
+
+	// Current version section is not completely consumed
+	++itemIndex;
+
+	const MutationRef::Type type = static_cast<MutationRef::Type>(
+	    *(deserializer.peekBytes(sizeof(Subsequence) + sizeof(MutationRef::Type)) + sizeof(Subsequence)));
```

Comment:
I think only peeking `sizeof(MutationRef::Type))` bytes is enough.

Reply:
It is necessary to skip the `Subsequence` as the data is serialized into format
```
| Subsequence | MutationRef::Type | MutationRef.param1 | MutationRef.param2 |
```

Reply:
You are right. My bad.